File: fs/before_checksum/4__median-of-two-sorted-arrays/cpp/median-of-two-sorted-arrays_benchmark_v2.cpp
#include <iostream>
#include <vector>
#include <functional>
#include <algorithm>
#include <limits>
using namespace std;

// Time:  O(log(min(m, n)))
// Space: O(1)
class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        // binary_search returns the smallest index i in [left, right] such that check(i) is true
        auto binary_search = [&](int left, int right, const function<bool(int)>& check) -> int {
            while (left <= right) {
                int mid = left + (right - left) / 2;
                if (check(mid)) {
                    right = mid - 1;
                } else {
                    left = mid + 1;
                }
            }
            return left;
        };

        auto getKth = [&](const vector<int>& A1, const vector<int>& B1, int k) -> int {
            const vector<int>* A = &A1;
            const vector<int>* B = &B1;
            int m = static_cast<int>(A->size());
            int n = static_cast<int>(B->size());
            if (m > n) {
                swap(A, B);
                swap(m, n);
            }
            int left = max(k - n, 0);
            int right = min(m, k) - 1;
            auto check = [&](int i) -> bool {
                int j = k - 1 - i;
                return (*A)[i] >= (*B)[j];
            };
            int i = binary_search(left, right, check);
            long long leftA = (i - 1 >= 0) ? static_cast<long long>((*A)[i - 1]) : numeric_limits<long long>::min();
            int j = k - 1 - i;
            long long leftB = (j >= 0) ? static_cast<long long>((*B)[j]) : numeric_limits<long long>::min();
            long long ans = max(leftA, leftB);
            return static_cast<int>(ans);
        };

        int len1 = static_cast<int>(nums1.size());
        int len2 = static_cast<int>(nums2.size());
        int total = len1 + len2;
        if (total % 2 == 1) {
            return static_cast<double>(getKth(nums1, nums2, total / 2 + 1));
        } else {
            int a = getKth(nums1, nums2, total / 2);
            int b = getKth(nums1, nums2, total / 2 + 1);
            return (static_cast<double>(a) + static_cast<double>(b)) * 0.5;
        }
    }
};

// Time:  O(log(max(m, n)) * log(max_val - min_val))
// Space: O(1)
// Generic solution.

int main() {
    vector<vector<vector<int>>> tests = {
        { {1, 3}, {2} },
        { {1, 2}, {3, 4} },
        { {0, 0, 0, 0}, {1, 2, 3} },
        { {-5, -3, -1}, {-2, 0, 2, 4} },
        { {1, 1, 1, 1, 1}, {1, 1, 2, 2, 2} },
        { {-2147483647, -1000000000, 0, 1000000000, 2147483647}, {-999999999, -1, 1, 999999999} },
        { {1, 3, 5, 7, 9, 11, 13, 15}, {2} },
        { {10}, {20} },
        { {100, 200, 300}, {1, 2, 3, 4, 5, 6, 7, 8, 9, 10} },
        { {1}, {2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15} }
    };

    Solution sol;
    volatile int sink = 0;

    const int iterations = 1;
    for (int iter = 0; iter < iterations; ++iter) {
        int checksum = 0;
        for (int t = 0; t < static_cast<int>(tests.size()); ++t) {
            vector<int>& a = tests[t][0];
            vector<int>& b = tests[t][1];
            double res = sol.findMedianSortedArrays(a, b);
            checksum += static_cast<int>(res);
        }
        sink += checksum;
        

    } std::cout << sink << std::endl;
    return 0;
}

File: fs/py_before_checksum/4__median-of-two-sorted-arrays/median-of-two-sorted-arrays_benchmark_v2.py
# Time:  O(log(min(m, n)))
# Space: O(1)

class Solution(object):
    def findMedianSortedArrays(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: float
        """
        def binary_search(left, right, check):
            while left <= right:
                mid = left+(right-left)//2
                if check(mid):
                    right = mid-1
                else:
                    left = mid+1
            return left

        def getKth(A, B, k):
            m, n = len(A), len(B)
            if m > n:
                m, n = n, m
                A, B = B, A
            i = binary_search(max(k-n, 0), min(m, k)-1, lambda i: A[i] >= B[k-1-i])
            return max(A[i-1] if i-1 >= 0 else float("-inf"), B[k-1-i] if k-1-i >= 0 else float("-inf"))

        len1, len2 = len(nums1), len(nums2)
        if (len1+len2) % 2 == 1:
            return getKth(nums1, nums2, (len1+len2)//2+1)
        else:
            return (getKth(nums1, nums2, (len1+len2)//2)+getKth(nums1, nums2, (len1+len2)//2+1))*0.5    


# Time:  O(log(max(m, n)) * log(max_val - min_val))
# Space: O(1)
# Generic solution.
class Solution_Generic(object):
    def findMedianSortedArrays(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: float
        """
        def binary_search(left, right, check):
            while left <= right:
                mid = left+(right-left)//2
                if check(mid):
                    right = mid-1
                else:
                    left = mid+1
            return left

        def getKth(arrays, k):
            def check(num):
                # count the number of values which are less or equal to num
                return sum(binary_search(0, len(arr)-1, lambda x: arr[x] > num) for arr in arrays) >= k
    
            return binary_search(min(arr[0] for arr in arrays if arr), max(arr[-1] for arr in arrays if arr), check)

        array = [nums1, nums2]
        total = sum(len(nums) for nums in array)
        if total % 2 == 1:
            return getKth(array, total//2+1)
        else:
            return (getKth(array, total//2)+getKth(array, total//2+1))*0.5


if __name__ == "__main__":
    tests = [
        [[1, 3], [2]],
        [[1, 2], [3, 4]],
        [[0, 0, 0, 0], [1, 2, 3]],
        [[-5, -3, -1], [-2, 0, 2, 4]],
        [[1, 1, 1, 1, 1], [1, 1, 2, 2, 2]],
        [[-2147483647, -1000000000, 0, 1000000000, 2147483647], [-999999999, -1, 1, 999999999]],
        [[1, 3, 5, 7, 9, 11, 13, 15], [2]],
        [[10], [20]],
        [[100, 200, 300], [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]],
        [[1], [2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]]
    ]

    sol = Solution()
    sink = 0

    iterations = 1
    for iter in xrange(iterations):
        checksum = 0
        for t in xrange(len(tests)):
            a = tests[t][0]
            b = tests[t][1]
            res = sol.findMedianSortedArrays(a, b)
            checksum += int(res)
        sink += checksum
    print(sink)